src/shared/services/shortcut-action.service.ts

```diff
@@ -0,0 +1,200 @@
+      path[path.length - 2] = ((elemIndexInPath + direction) < values.size &&
+                              (elemIndexInPath + direction) >= 0) ? elemIndexInPath + direction :
+                              values.size - Math.abs((elemIndexInPath + direction));
+      this.domUtilService.focusAndSelectFirstInputChildById(path.join('.'));
+    }
+  }
+
+  navigateLeftAction(path: Array<any>): void {
+    this.navigateRightLeftAction(path, -1);
+  }
+
+  navigateRightAction(path: Array<any>): void {
+    this.navigateRightLeftAction(path, 1);
+  }
+
+  private navigateRightLeftAction(path: Array<any>, direction: number) {
+    let tabIndex = this.tabIndexService.getElemTabIndex(path.join('.'));
+    if (tabIndex >= 0) {
+      this.domUtilService.focusRightOrLeftCell(tabIndex, direction);
+    }
+  }
+
+  copyRowAction(path: Array<any>) {
+    let _path = this.pathUtilService.getNearestOrRootArrayParentInPath(path, false);
+    this.copyCurrentRowBelow(path, _path);
+  }
+
+  copyItemAction(path: Array<any>) {
+    let _path = this.pathUtilService.getNearestOrRootArrayParentInPath(path, true);
+    this.copyCurrentSchemaItemBelow(path, _path);
+  }
+
+  private copyCurrentRowBelow(originalPath: Array<any>, arrayParentPath: Array<any>) {
+    let elemIndex = this.pathUtilService.getElementIndexInReversePath(originalPath);
+    let valuesList = this.jsonStoreService.getIn(arrayParentPath) || List.of([]);
+    let newValue = valuesList.get(elemIndex);
+    newValue = newValue.set(originalPath[originalPath.length - 1], '');
+    this.jsonStoreService.setIn(arrayParentPath, valuesList.insert(elemIndex + 1, newValue));
+    arrayParentPath.push(elemIndex + 1);
+    setTimeout(() => {
+      this.updateTabIndexesAndFocusElementInPath(arrayParentPath.join('.').concat('.' + originalPath[originalPath.length - 1]));
+    });
+  }
+
+  private copyCurrentSchemaItemBelow(originalPath: Array<any>, arrayParentPath: Array<any>) {
+    let elemIndex = this.pathUtilService.getElementIndexInPath(originalPath);
+    let valuesList = this.jsonStoreService.getIn(arrayParentPath) || List.of([]);
+    let newValue = valuesList.get(elemIndex);
+    this.jsonStoreService.setIn(arrayParentPath, valuesList.insert(elemIndex + 1, newValue));
+    arrayParentPath.push(elemIndex + 1);
+    setTimeout(() => {
+      this.updateTabIndexesAndFocusElementInPath(arrayParentPath.join('.'));
+    });
+  }
+
+  private updateTabIndexesAndFocusElementInPath(path: string) {
+    this.tabIndexService.insertElemTabIndex(path);
+    this.domUtilService.flashElementById(path);
+    this.domUtilService.focusAndSelectFirstInputChildById(path);
+  }
+
+  generateShortcutAction(action: string) {
+
+    return (event: KeyboardEvent) => {
+      event.preventDefault();
+      let eventTarget = event.target as HTMLInputElement;
+      this[action](this.tabIndexService.getPathFromTabIndex(eventTarget.tabIndex));
+      return false;
```

Comment:
I think first one is disable the default action, and the return is for bubbling, since they might have pressed the shortcut for longer time.